### ISSUE TITLE
csv/file_utils: Write files atomically

### ISSUE THREAD

**Issue by papr** [pupil-labs/pupil#1363]
Follow up issue for #1304 , solution proposal by @windsooon: https://github.com/pupil-labs/pupil/issues/1304#issuecomment-434681050

**Comment by ckbaumann** [comment]
I would suggest to have a class for this that can be used with python `with` statement, e.g.
```
with atomic_open(meta_info_path,'w',newline='',encoding='utf-8') as csvfile: 
     csv_utils.write_key_value_file(csvfile,meta_info) 
```
Would be a nice opportunity to practice writing a functional test for concurrency issues.

**Comment by Windsooon** [comment]
Yes, the easy fix doesn't work on Windows, I found a repo https://github.com/benediktschmitt/py-filelock which implement using threading lock which should work. However, I haven't read the source code so I'm not sure it works as expected.

**Comment by ckbaumann** [comment]
Thanks for the research!

There seem to be different kinds of file locking libraries:
- Try to lock files on system level (i.e. the OS should enforce that noone else can change the file under lock), e.g. [`portalocker`](https://pypi.org/project/portalocker/) or this [custom code](https://stackoverflow.com/a/46407326)
- Try to "lock" a file (e.g. `example.txt`) by creating a lock file (e.g. `example.txt.lock`) such that other processes know that they should not change the file. As far as I understand, this is implemented in `py-filelock`, but I'm not quite sure, it also seems to lock the file at os level (it offers [system independend soft locks and non-soft locks](https://github.com/benediktschmitt/py-filelock#filelock-vs-softfilelock))
- Use locks inside the program only that make sure that different processes of the same program do not access the file simultaneously, e.g. [`fasteners`](https://pypi.org/project/fasteners/) (however, this seems to only provide locks between different _processes_, not different _threads_ of the same process.

In conclusion, there seem to be a variety of different options with no clear best approach standing out. We would need to setup a test program and carefully test different libraries under different operating systems.

Me and @papr think that this is currently not urgent and that there are other more pressing problems. Hence, we will not fix this immediately.

**Comment by papr** [comment]
Instead of ensuring files are written atomically, we have removed any code that would cause a race codnition in the first place.
